R/__ddl_by_AC.R

```diff
@@ -0,0 +1,237 @@
+    call <- as.list(call)
+    call[["inputId"]] <- call("ns", call[["inputId"]])
+    as.call(call)
+  })
+
+
+  ui <- function(id) {
+    ns <- NS(id)
+    div(
+      class = "connector-input",
+      lapply(args, eval, envir = environment())
+    )
+  }
+
+  if (missing(mask)) mask <- list()
+  tracked_request <- with_substitution(on_submit, mask)
+  server <- function(id) {
+    moduleServer(id, function(input, output, session) {
+      result <- eventReactive(input[["submit"]], {
+        inputs <- sapply(setdiff(inputIds, "submit"), function(x) input[[x]], simplify = FALSE)
+        do.call(tracked_request, list(inputs))
+      })
+      result
+    })
+  }
+
+  list(
+    ui = ui,
+    server = server
+  )
+}
+
+
+#' wrapper for `on_submit` functions
+#'
+#' Wrap a function that makes some assignments in its body to return a `tdata` object with optionally masked code.
+#'
+#' Code found in the body of `fun` will be run in order to obtain the desired data sets.
+#' References to `input$<name>` will be substituted with input values of the accompanying `shiny` module
+#' for the purposes of code execution. If `mask` is provided, those references will be substituted with mask values
+#' for the purposes of storing code.
+#'
+#' @param fun a function that takes exactly one argument, `input`, which is a named list
+#' @param mask optional named list to specify code masking; see `input_template` for details
+#'
+#' @return
+#' A `tdata` object containing variables that were created in the body of `fun`
+#' and the entirety of the body of `fun` in the `@code` slot.
+#'
+#' @keywords internal
+with_substitution <- function(fun, mask) {
+  checkmate::assert_true(
+    identical(names(formals(fun)), "input"),
+    .var.name = "'on_submit' function only takes 'input' argument"
+  )
+  checkmate::assert_list(mask, names = "unique")
+  function(...) {
+    # Get input values from call arguments.
+    call_args <- as.list(match.call(fun))$input
+    checkmate::assert_list(call_args, names = "unique", .var.name = "input")
+    # Add non-masked arguments to mask.
+    mask <- c(mask, call_args)
+    mask <- mask[!duplicated(names(mask))]
+
+    # Extract function body as list of calls.
+    fun_body <- body(fun)
+    code <-
+      if (is.expression(fun_body)) {
+        as.list(fun_body)
+      } else if (is.call(fun_body)) {
+        if (identical(as.list(fun_body)[[1L]], as.symbol("{"))) {
+          as.list(fun_body)[-1L]
+        } else {
+          list(fun_body)
+        }
+      } else if (is.name(fun_body)) {
+        fun_body
+      } else {
+        stop("with_substitution: don't know ho to handle this kind of function body")
+      }
+
+    # Convert calls to strings and substitute argument references by bquote references.
+    code_strings <- vapply(code, deparse1, character(1L))
+    code_strings <- gsub("(input\\$)(\\w+)", "\\.(\\2\\)", code_strings)
+    code_strings <- gsub("(input\\[\\[\")(\\w+)(\"\\]\\])", "\\.(\\2\\)", code_strings)
```

Comment:
@chlebowa I would make sure those regexes work on all platforms. Looks like system specific?

Reply:
As far as I know this is just R-specific, it looks a little weird because every slash has to be escaped. `\w` and backtracing are widely recognized.

Reply:
coolio

Reply:
There may be some edge cases where users use `purrr::plucK(input, "something")` or something else.

Reply:
True, `get, for instance. I'm not sure we have to worry about it, considering people mosty use `$`.

Now that I think about it, some NSE tricks would probably go unnoticed :thinking: